src/mesh_routines.f90

```diff
@@ -1853,14 +1858,22 @@ SUBROUTINE DECOMP_TOPOLOGY_ELEM_ADJACENT_ELEM_CALCULATE(TOPOLOGY,ERR,ERROR,*)
                                 candidate_el=DOMAIN_NODES%NODES(node_idx)%SURROUNDING_ELEMENTS(surrounding_el_idx)
                                 IF(candidate_el/=ne) THEN
                                   candidate_idx=candidate_idx+1
-                                  SURROUNDING_ELEMENTS(candidate_idx)=candidate_el
+                                  CALL LIST_ITEM_ADD(SURROUNDING_ELEMENTS,candidate_el,ERR,ERROR,*999)
                                 ENDIF
                               ENDDO
                             ENDDO !face_node_idx
                             DO idx=1,candidate_idx
-                              ne1=SURROUNDING_ELEMENTS(idx)
+                              CALL LIST_ITEM_GET(SURROUNDING_ELEMENTS,idx,ne1,ERR,ERROR,*999)
                               !In a 2D mesh, match 2 nodes, in a 3D mesh, match 3 nodes (line/face)
-                              IF(COUNT(SURROUNDING_ELEMENTS(1:candidate_idx)==ne1)>=BASIS%NUMBER_OF_XI) THEN
```

Comment:
We try to get the data structures as encapsulated as possible (i.e., not directly access the type components directly). What is probably better is that you do a list detach and destroy call that returns you an integer array and the number of items in the list. This will also destroy the list (which is not currently done and so there will be a memory leak). You will then have to deallocate the integer array when you have finished with it and when there is an error condition at the end of the subroutine.